connmanager/connmanager.go

```diff
@@ -0,0 +1,139 @@
+
+	stop                   uint32
+	connectionRequestsLock sync.Mutex
+}
+
+// New instantiates a new instance of a ConnectionManager
+func New(netAdapter *netadapter.NetAdapter, addressManager *addrmgr.AddrManager) (*ConnectionManager, error) {
+	c := &ConnectionManager{
+		netAdapter:                netAdapter,
+		addressManager:            addressManager,
+		activeConnectionRequests:  map[string]*connectionRequest{},
+		pendingConnectionRequests: map[string]*connectionRequest{},
+		activeOutgoing:            map[string]struct{}{},
+		activeIncoming:            map[string]struct{}{},
+	}
+
+	cfg := config.ActiveConfig()
+	connectPeers := cfg.AddPeers
+	if len(cfg.ConnectPeers) > 0 {
+		connectPeers = cfg.ConnectPeers
+	}
+
+	c.maxIncoming = cfg.MaxInboundPeers
+	c.targetOutgoing = cfg.TargetOutboundPeers
+
+	for _, connectPeer := range connectPeers {
+		c.pendingConnectionRequests[connectPeer] = &connectionRequest{
+			address:     connectPeer,
+			isPermanent: true,
+		}
+	}
+
+	return c, nil
+}
+
+// Start begins the operation of the ConnectionManager
+func (c *ConnectionManager) Start() {
+	cfg := config.ActiveConfig()
+	if !cfg.DisableDNSSeed {
+		dnsseed.SeedFromDNS(cfg.NetParams(), wire.SFNodeNetwork, false, nil,
+			config.ActiveConfig().Lookup, func(addrs []*wire.NetAddress) {
+				// Kaspad uses a lookup of the dns seeder here. Since seeder returns
+				// IPs of nodes and not its own IP, we can not know real IP of
+				// source. So we'll take first returned address as source.
+				c.addressManager.AddAddresses(addrs, addrs[0], nil)
+			})
+	}
+
+	spawn(c.connectionsLoop)
+}
+
+// Stop halts the operation of the ConnectionManager
+func (c *ConnectionManager) Stop() {
+	atomic.StoreUint32(&c.stop, 1)
+
+	for _, connection := range c.netAdapter.Connections() {
+		_ = connection.Disconnect() // Ignore errors since connection might be in the midst of disconnecting
+	}
+}
+
+const connectionsLoopInterval = 30 * time.Second
+
+func (c *ConnectionManager) initiateConnection(address string) error {
+	log.Infof("Connecting to %s", address)
+	_, err := c.netAdapter.Connect(address)
+	if err != nil {
+		log.Infof("Couldn't connect to %s: %s", address, err)
+	}
+	return err
+}
+
+func (c *ConnectionManager) connectionsLoop() {
+	for atomic.LoadUint32(&c.stop) == 0 {
+		connections := c.netAdapter.Connections()
+
+		connSet := convertToSet(connections)
+
+		c.checkConnectionRequests(connSet)
+
+		c.checkOutgoingConnections(connSet)
+
+		c.checkIncomingConnections(connSet)
+
+		<-time.Tick(connectionsLoopInterval)
+	}
+}
+
+// checkIncomingConnections makes sure there's no more then maxIncoming incoming connections
+// if there are - it randomly disconnects enough to go below that number
+func (c *ConnectionManager) checkIncomingConnections(connSet connectionSet) {
+	if len(connSet) <= c.maxIncoming {
+		return
+	}
+
+	// randomly disconnect nodes until the number of incoming connections is smaller the maxIncoming
```

Comment:
This isn't so easy, since connectionManager doesn't get pings whenever somebody connects, and can't decide to disconnect immidiately.
Deferring solving this to a later date.

Reply:
So please open a ticket for this